ath79: add support for Comfast CF-WR752AC v1

rhamp opened the pull request:
- Qualcomm QCA9531 + QCA9886
- dual band, antenna 2*3dBi
- Output power 50mW (17dBm)
- 1x 10/100 Mbps LAN RJ45
- 128MB RAM / 16 MB FLASH
- 3 LEDs (red/green/blue)
  incorporated in
  "color wheel reset switch"
- UART 115200 8N1

Flashing instructions:

 The U-boot bootloader contains a recovery HTTP server
 to upload the  firmware. Push the reset button while powering the
 device on and keep it pressed for ~10 seconds. The device's LEDs will
 blink several times and the recovery page will be at
 http://192.168.1.1; use it to upload the sysupgrade image.

 Alternatively, the original firmware is based on OpenWrt so a
 sysupgrade image can be installed via the stock web GUI. Settings from
 the original firmware will be saved and restored on the new one, so a
 factory reset will be needed. To do so, once the new firmware is flashed,
 enter into failsafe mode by pressing the reset button several times during
 the boot process, until it starts flashing. Once in failsafe mode, perform
 a factory reset as usual.

Signed-off-by: Roman Hampel <rhamp@arcor.de>

adrianschmutzler commented:
This is a duplicate of https://github.com/openwrt/openwrt/pull/2652.

Since the old one seems to be discontinued somehow, though, I won't close this one just because of that.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+
+/ {
+	compatible = "comfast,cf-wr752ac", "qca,qca9531";
+	model = "COMFAST CF-WR752AC";
```
If this is a v1, please include that in all the names:
```
	compatible = "comfast,cf-wr752ac-v1", "qca,qca9531";
	model = "COMFAST CF-WR752AC v1";
```
Same for file names, imake Makefile, board.d etc.

rhamp replied:
I didn't use v1 since there is no mention on the device itself, but since it seems to be the good  usance, no problem.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+		serial0 = &uart;
+		led-boot = &lan;
+		led-failsafe = &lan;
+		led-upgrade = &lan;
```
I'm not a fan of using lan and wlan for indication here.

rhamp replied:
I don't really like it too, but as far as I can see, Comfast uses the LAN led too for signaling events. There are no other vacant leds available.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins &led_lan &led_wlan2g>;
+
+		lan: lan {
```
If you insist on using this for led aliases, please prefix the _DTS label_ with "led_", so this becomes:
`led_lan: lan {`

rhamp replied:
OK, if you don't mind that alias names get longer, I can do that.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+		pinctrl-0 = <&jtag_disable_pins &led_lan &led_wlan2g>;
+
+		lan: lan {
+			label = "cf-wr752ac:green:lan";
```
-> cf-wr752ac-v1

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+			label = "cf-wr752ac:green:lan";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "netdev";
+			trigger-sources = <&eth0>;
```
I do not think this netdev trigger is working. You will have to use 01_leds.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "netdev";
+			trigger-sources = <&eth0>;
+			default-state = "on";
```
No. This is an ethernet LED.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+		wlan2g {
+			label = "cf-wr752ac:red:wlan2g";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
Remove this line, it's default.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+			linux,default-trigger = "phy1tpt";
+			panic-indicator;
```
I do not like that on a wireless LED either.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+			default-state = "off";
+			linux,default-trigger = "phy0tpt";
+		};
+
```
Remove this empty line.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+	};
+};
+
+
```
No double empty lines.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+
+
+&pinmux {
+	led_lan: pinmux_lan {
```
-> `pinmux_lan_led: lan_led` or `pinmux_led_lan: led_lan`

rhamp replied:
Since I don't like long aliases, can we meet in the middle with pinmux_lan ? ;-)

adrianschmutzler replied:
Well, I like accurate aliases, and you are not changing lan here, but the lan led?

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+		pinctrl-single,bits = <0x0c 0x0 0xff000000>;
+	};
+
+	led_wlan2g: pinmux_wlan2g {
```
Change as indicated above

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
```
Add empty line after status.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "winbond,w25q128", "jedec,spi-nor" ;
```
Remove space before ;

rhamp replied:
;-) I left school a long time ago, that makes me remember the time.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+};
+
+&eth0 {
+	status = "okay";
```
Empty line after status.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+};
+
+&eth1 {
+	// "Disable" unpopulated GMAC1
```
I do not think this comment is needed.

rhamp replied:
Are you sure that everyone knows that eth1 has to be disabled this unorthodox way, since there is no physical eth1 present, and the whole mess with eth0 <-> eth1 name switching ...
Precedent case is in file "qca9531_glinet_gl-ar300m-lite.dts".

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+&eth1 {
+	// "Disable" unpopulated GMAC1
+	compatible = "syscon", "simple-mfd";
+	mtd-mac-address = <&art 0x6>;
```
You do not need a MAC address for a disabled interface.

rhamp replied:
It's just to be thorough , since it is present in Comfast art partition. But can remove it for file to  be more slim.

adrianschmutzler replied:
MAC assignment should match the assignment in stock firmware. How will you find out the matching address for a non-existent interface?

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+};
+
+&wmac {
+	status = "okay";
```
Empty line after status.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+&wmac {
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x1002>;
```
This line can be removed, as it will be done by mtd-cal-data automatically.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -80,6 +80,11 @@ comfast,cf-e560ac)
 	ucidef_set_led_switch "lan4" "LAN4" "$boardname:blue:lan4" "switch0" "0x10"
 	;;
+comfast,cf-wr752ac)
+        ucidef_set_led_netdev "lan" "LAN" "$boardname:green:lan" "eth0"
```
Keep this line, as trigger won't work here, and remove the WLAN lines, as this is done by trigger.
Please check whether you can merge with another case then.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -137,6 +137,15 @@ case "$FIRMWARE" in
+		ln -sf /lib/firmware/ath10k/pre-cal-pci-0000\:00\:00.0.bin \
+			/lib/firmware/ath10k/QCA9888/hw2.0/board.bin
+		ln -sf /lib/firmware/ath10k/QCA9888/hw2.0/firmware-5.bin \
+			/lib/firmware/ath10k/QCA9888/hw2.0/ct-firmware-5.bin
```
Shouldn't one be enough?

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -137,6 +137,15 @@ case "$FIRMWARE" in
+		ln -sf /lib/firmware/ath10k/QCA9888/hw2.0/firmware-5.bin \
+			/lib/firmware/ath10k/QCA9888/hw2.0/ct-firmware-5.bin
+		board2file="/lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin"
+		[ -e $board2file ] && rm $board2file
```
To me it looks like you can just use what's already there for cf-e560ac?

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -300,6 +300,16 @@ define Device/comfast_cf-e5
+define Device/comfast_cf-wr752ac
+  SOC := qca9531
+  DEVICE_VENDOR := COMFAST
+  DEVICE_MODEL := CF-WR752AC
```
DEVICE_VARIANT := v1

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -300,6 +300,16 @@ define Device/comfast_cf-e5
+  DEVICE_VENDOR := COMFAST
+  DEVICE_MODEL := CF-WR752AC
+  DEVICE_PACKAGES := kmod-leds-gpio kmod-usb2 kmod-ath10k-ct ath10k-firmware-qca9888-ct \
+	-uboot-envtools swconfig
```
Isn't swconfig there by default? Or did you actually intend to remove it?

adrianschmutzler replied:
Does this have USB?

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -300,6 +300,16 @@ define Device/comfast_cf-e5
 endef
 TARGET_DEVICES += comfast_cf-e5
 
+define Device/comfast_cf-wr752ac
```
Alphabetic sorting!

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot: partition@0 {
```
The only node label you need should be art. So, you can remove the uboot:, firmware: ...

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+			firmware: partition@20000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x020000 0xfd0000>;
```
The older PR has a configs partition at 0xfe0000 ...

rhamp replied:
There is no need for a config partition with default settings of stock fw. More space for openwrt. It is like this  on many Comfast ath79 devices.

adrianschmutzler replied:
So, is the config partition recreated by the original firmware then? Or are you really just throwing away the option to flash back vendor OS for 64k on a 16 MB flash?

Which ath79 devices are you talking about?

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac.dts:
```diff
@@ -0,0 +1,148 @@
+&eth0 {
+	status = "okay";
+	phy-handle = <&swphy4>;
+	mtd-mac-address = <&art 0x0>;
```
The old PR sets this up differently.

rhamp replied:
The 9531 has a build-in 5 port switch, manufacturers can connect physical eth at phy0 or phy4. This device has it at the latter. I don't know how it shall work without it.

adrianschmutzler commented:
Hi, despite my comments above, note that you have entirely different LED setup than in the old PR. Please check and adjust that if necessary.
In addition, we found out that this device has quite a strange MAC address assignment. Please report which valid addresses you find in art partition and report the MAC address assignment on vendor firmware as well.
Further reading: https://openwrt.org/docs/guide-developer/mac.address

joaohcca commented:
Hi, I'm the author of  [2652](https://github.com/openwrt/openwrt/pull/2652) 
The original firmware uses red and green led as a signal indicator, although is not crucial, why did you decide to change their behavior? 

I'm also interested in the MAC address assignment, I have a few COMFAST devices which doesn't have a consistency in this matter.

Do you have your device's original firmware bootlog? Mine have some partition definition sightly different. I'm attaching here for reference. 

@adrianschmutzler this PR seems much closer to being accepted than mine, so feel free to close mine, if you deem necessary.

`## Booting image at 9f020000 ...                                                
   Image Name:   MIPS OpenWrt Linux-3.10.44                                     
   Created:      2018-01-05   8:59:30 UTC                                       
   Image Type:   MIPS Linux Kernel Image (lzma compressed)                      
   Data Size:    1124162 Bytes =  1.1 MB                                        
   Load Address: 80060000                                                       
   Entry Point:  80060000                                                       
   Verifying Checksum at 0x9f020040 ...OK                                       
   Uncompressing Kernel Image ... OK                                            
No initrd                                                                       
 Transferring control to Linux (at address 80060000) ...                      
Giving linux memsize in bytes, 134217728                                     
                                                                                
Starting kernel ...                                                             
                                                                                
[    0.000000] Linux version 3.10.44 (wangqiuqiang@oracle.tecmint.com) (gcc ver8
[    0.000000] bootconsole [early0] enabled                                     
[    0.000000] CPU revision is: 00019374 (MIPS 24Kc)                            
[    0.000000] SoC: Qualcomm Atheros QCA9531 ver 2 rev 0                        
[    0.000000] Clocks: CPU:650.000MHz, DDR:391.650MHz, AHB:216.666MHz, Ref:25.0z
[    0.000000] Determined physical RAM map:                                     
[    0.000000]  memory: 08000000 @ 00000000 (usable)                            
[    0.000000] Initrd not found or empty - disabling initrd                     
[    0.000000] Zone ranges:                                                     
[    0.000000]   Normal   [mem 0x00000000-0x07ffffff]                           
[    0.000000] Movable zone start for each node                                 
[    0.000000] Early memory node ranges                                         
[    0.000000]   node   0: [mem 0x00000000-0x07ffffff]                          
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.  
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32s
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pa2
[    0.000000] Kernel command line:  board=COMFAST-CF-WR752AC console=ttyS0,115d
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)              
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)   
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)     
[    0.000000] Writing ErrCtl register=00000000                                 
[    0.000000] Readback ErrCtl register=00000000                                
[    0.000000] Memory: 125796k/131072k available (2425k kernel code, 5276k rese)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1       
[    0.000000] NR_IRQS:51                                                       
[    0.000000] Calibrating delay loop... 432.53 BogoMIPS (lpj=2162688)          
[    0.050000] pid_max: default: 32768 minimum: 301                             
[    0.050000] Mount-cache hash table entries: 512                              
[    0.060000] NET: Registered protocol family 16                               
[    0.070000] MIPS: machine is COMFAST CF-WR752AC                              
[    0.510000] registering PCI controller with io_map_base unset                
[    0.520000] bio: create slab <bio-0> at 0                                    
[    0.530000] PCI host bridge to bus 0000:00                                   
[    0.530000] pci_bus 0000:00: root bus resource [mem 0x10000000-0x11ffffff]   
[    0.540000] pci_bus 0000:00: root bus resource [io  0x0000]                  
[    0.540000] pci_bus 0000:00: No busn resource found for root bus, will use []
[    0.550000] pci 0000:00:00.0: invalid calibration data                       
[    0.550000] pci 0000:00:00.0: BAR 0: assigned [mem 0x10000000-0x101fffff 64b]
[    0.560000] pci 0000:00:00.0: using irq 40 for pin 1                         
[    0.560000] Switching to clocksource MIPS                                    
[    0.570000] NET: Registered protocol family 2                                
[    0.570000] TCP established hash table entries: 1024 (order: 1, 8192 bytes)  
[    0.580000] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)         
[    0.580000] TCP: Hash tables configured (established 1024 bind 1024)         
[    0.590000] TCP: reno registered                                             
[    0.590000] UDP hash table entries: 256 (order: 0, 4096 bytes)               
[    0.600000] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)          
[    0.600000] NET: Registered protocol family 1                                
[    0.620000] squashfs: version 4.0 (2009/01/31) Phillip Lougher               
[    0.630000] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORI.
[    0.640000] msgmni has been set to 245                                       
[    0.650000] io scheduler noop registered                                     
[    0.650000] io scheduler deadline registered (default)                       
[    0.650000] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled         
[    0.680000] serial8250.0: ttyS0 at MMIO 0x18020000 (irq = 11) is a 16550A    
[    0.690000] console [ttyS0] enabled, bootconsole disabled                    
[    0.690000] console [ttyS0] enabled, bootconsole disabled                    
[    0.700000] ath79-spi ath79-spi: master is unqueued, this is deprecated      
[    0.710000] m25p80 spi0.0: found w25q128, expected m25p80                    
[    0.720000] m25p80 spi0.0: w25q128 (16384 Kbytes)                            
**[    0.720000] 7 cmdlinepart partitions found on MTD device spi0.0              
[    0.730000] Creating 7 MTD partitions on "spi0.0":                           
[    0.730000] 0x000000000000-0x000000010000 : "u-boot"                         
[    0.740000] 0x000000010000-0x000000020000 : "art"                            
[    0.750000] 0x000000020000-0x0000001a0000 : "kernel"                         
[    0.750000] 0x0000001a0000-0x000000fe0000 : "rootfs"                         
[    0.760000] mtd: device 3 (rootfs) set to be root filesystem                 
[    0.770000] 1 squashfs-split partitions found on MTD device rootfs           
[    0.770000] 0x000000780000-0x000000fe0000 : "rootfs_data"                    
[    0.780000] 0x000000fe0000-0x000000ff0000 : "configs"                        
[    0.790000] 0x000000ff0000-0x000001000000 : "nvram"                          
[    0.800000] 0x000000020000-0x000000fe0000 : "firmware"**                       
[    0.820000] libphy: ag71xx_mdio: probed                                      
[    0.830000] eth0: Atheros AG71xx at 0xba000000, irq 5                        
[    1.390000] ag71xx-mdio.1: Found an AR934X built-in switch                   
[    1.430000] eth1: Atheros AG71xx at 0xb9000000, irq 4                        
[    1.990000] ag71xx ag71xx.0: connected to PHY at ag71xx-mdio.1:04 [uid=004dd]
[    2.000000] TCP: cubic registered                                            
[    2.000000] NET: Registered protocol family 17                               
[    2.010000] Bridge firewalling registered                                    
[    2.010000] 8021q: 802.1Q VLAN Support v1.8                                  
[    2.020000] VFS: Mounted root (squashfs filesystem) readonly on device 31:3. 
[    2.030000] Freeing unused kernel memory: 276K (8036b000 - 803b0000)         
[    2.990000] init: Console is alive                                           
[    2.990000] init: - watchdog -                                               
[    5.000000] init: - preinit -                                                
Press the [f] key and hit [enter] to enter failsafe mode                        
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level    
[    9.660000] jffs2: notice: (372) jffs2_build_xattr_subsystem: complete build.
[    9.680000] mount_root: switching to jffs2 overlay                           
2116+0 records in                                                               
2116+0 records out                                                              
12064+0 records in                                                              
12064+0 records out                                                             
[   10.220000] procd: - early -                                                 
[   10.230000] procd: - watchdog -                                              
[   10.920000] procd: - ubus -                                                  
[   11.930000] procd: - init -                                                  
Please press Enter to activate this console.                                    
[   13.850000] natsemi dp8381x driver, version 2.1, Sept 11, 2006               
[   13.850000]   originally by Donald Becker <becker@scyld.com>                 
[   13.850000]   2.4.x kernel port by Jeff Garzik, Tjeerd Mulder                
[   13.880000] gre: GRE over IPv4 demultiplexor driver                          
[   13.900000] ip_gre: GRE over IPv4 tunneling driver                           
[   13.910000] i2c /dev entries driver                                          
[   13.920000] Loading modules backported from Linux version master-2014-05-22-a
[   13.930000] Backport generated by backports.git backports-20140320-37-g5c33d0
[   14.170000] PCI: Enabling device 0000:00:00.0 (0000 -> 0002)                 
[   14.180000] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode0
[   15.090000] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id0
[   15.100000] ath10k_pci 0000:00:00.0: kconfig debug 1 debugfs 1 tracing 1 dfs1
[   15.130000] ath10k_pci 0000:00:00.0: firmware ver 10.4-3.5.1-00014 api 5 feaa
[   15.520000] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id 0:0 crc32 9cddb9
[   17.310000] ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal otp m1
[   17.460000] u32 classifier                                                   
[   17.460000]     input device check on                                        
[   17.460000]     Actions configured                                           
[   17.480000] Mirror/redirect action on                                        
[   17.510000] nf_conntrack version 0.5.0 (1969 buckets, 7876 max)              
[   17.590000] Ebtables v2.0 registered                                         
[   17.600000] ip_tables: (C) 2000-2006 Netfilter Core Team                     
[   17.660000] Netfilter messages via NETLINK v0.30.                            
[   17.720000] xt_time: kernel timezone is -0000                                
[   17.750000] ctnetlink v0.93: registering with nfnetlink.                     
[   17.760000] PPP generic driver version 2.4.2                                 
[   17.770000] PPP MPPE Compression module registered                           
[   17.780000] NET: Registered protocol family 24                               
[   17.840000] ieee80211 phy1: Atheros AR9531 Rev:2 mem=0xb8100000, irq=47      
configs:do check configs                                                        
[   28.130000] device eth0 entered promiscuous mode                             
[   28.180000] device eth1 entered promiscuous mode                             
[   28.210000] br-wan: port 1(eth1) entered forwarding state                    
[   28.220000] br-wan: port 1(eth1) entered forwarding state                    
[   29.070000] br-wan: port 1(eth1) entered disabled state                      
[   32.160000] device wlan8 entered promiscuous mode                            
[   35.780000] device wlan0 entered promiscuous mode                            
[   41.530000] br-lan: port 2(wlan8) entered forwarding state                   
[   41.530000] br-lan: port 2(wlan8) entered forwarding state                   
[   43.530000] br-lan: port 2(wlan8) entered forwarding state                   
[   46.240000] br-lan: port 3(wlan0) entered forwarding state                   
[   46.240000] br-lan: port 3(wlan0) entered forwarding state                   
[   48.240000] br-lan: port 3(wlan0) entered forwarding state 
`

adrianschmutzler commented:
> The original firmware uses red and green led as a signal indicator, although is not crucial, why did you decide to change their behavior?

Note that both of you also assign the LEDs to different GPIOs. So maybe sort out LEDs and MAC addresses and partition layout first.

rhamp commented:
The ART partition contains several valid, consecutive MAC addresses at 
0x0 xx:xx:xx:xx:xx:6C   -> eth0
0x6 xx:xx:xx:xx:xx:6E
0x1002 xx:xx:xx:xx:xx:6D  -> wlan2g
0x5006 xx:xx:xx:xx:xx:6F  -> wlan5g
which are correctly read out and used at boot time with the present dts.

Regarding the leds, comfast has changed led usage with recent firmwares, that are now controlled by a proprietary crond script. I did not clone that script since it is their copyright.

adrianschmutzler commented:
> which are correctly read out and used at boot time with the present dts.

Note that "correctly" is defined as "what the vendor does, unless it's random there". So, please check what the vendor does.

adrianschmutzler commented:
BTW: To me, the MAC setup on this device seems to be similar to what CF-E560AC does:
https://github.com/openwrt/openwrt/pull/2718
https://github.com/openwrt/openwrt/commit/ae61d21ca30f3af2732f73f82a01bba7e4341459

joaohcca commented:
> BTW: To me, the MAC setup on this device seems to be similar to what CF-E560AC does:
> #2718
> [ae61d21](https://github.com/openwrt/openwrt/commit/ae61d21ca30f3af2732f73f82a01bba7e4341459)

I agree, based on my PR discussion the 5GHz MAC Adress is incremented by 8 too.

> For reference the original firmware had the folloing definitions:
xx:xx:xx:xx:dc:f8 lan port
xx:xx:xx:xx:dc:fa wifi 5GHz
xx:xx:xx:xx:dd:02 wifi 2.4GHz

Also, original firmware bootlog indicates that wlan0 and wlan8 are up

rhamp commented:
Matched MAC behavior to Comfast v1 fw from 2018.

rhamp commented:
Please check again.

adrianschmutzler commented:
Please squash into a single proper commit (git rebase -i ...) and then force push the merged commit to GitHub again (same branch, no new PR)

adrianschmutzler commented:
But besides, it looks like you haven't addressed several of my remarks.

rhamp commented:
> But besides, it looks like you haven't addressed several of my remarks.

If just for formatting stuff, I made the changes without replying here, Changed LED attribution back to factory (the blue led got very dim for my devs, that's why I switched at first). The MACs are corresponding to 2018' cf firmware (opposed to current 2020' cf fw, which extracts it from eeprom). Partitioning is OK. Did I forget anything?

adrianschmutzler commented:
> Info:
> All LEDs are working, blue is set to LAN, green to wlan2g and red to wlan5g,
> LAN port is set to eth0

This reads like it's your decision. Does it match the vendor behavior? If yes, you don't need this comment.

adrianschmutzler commented:
> cf-wr752ac-v1

I'd make the model name upper-case in commit title, and separate the v1:
"ath79: add support for Comfast CF-WR752AC v1"

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+		led_lan: lan {
+			label = "cf-wr752ac-v1:blue:lan";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "netdev";
```
Don't think this line does anything?

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+};
+
+&pinmux {
+	led_wlan2g_pin: pinmux_lan {
```
is it lan or wlan2g?
Despite, for the part after the colon, please include "led" in the name, but "pinmux" is not really required, as it's a child of the pinmux node anyway.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+		pinctrl-single,bits = <0x0c 0x0 0xff000000>;
+	};
+
+	led_wlan5g_pin: pinmux_wlan2g {
```
Non-matching name and label here as well.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+
+&pinmux {
+	led_wlan2g_pin: pinmux_lan {
+		pinctrl-single,bits = <0x0c 0x0 0xff000000>;
```
Consider 0x0c -> 0xc

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+	};
+
+	led_wlan5g_pin: pinmux_wlan2g {
+		pinctrl-single,bits = <0x08 0x0 0xff000000>;
```
Consider 0x08 -> 0x8

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "winbond,w25q128", "jedec,spi-nor";
```
Do you need the "winbond,w25q128" compatible?

rhamp replied:
used flash here is w25q128, 
precedent: qca9533_comfast_cf-e110n-v2.dts

adrianschmutzler replied:
And that's the only in entire ath79. "jedec,spi-nor" will provide auto-identification, so I will answer my question myself: You don't need it, please remove it and just keep "jedec,spi-nor".

rhamp replied:
OK. Removed.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,144 @@
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x020000 0xfd0000>;
+			};
```
So, what's the result of your research regarding the config partition?

rhamp replied:
If you return to manufacturer fw, it is restored with its default settings (as it should). No Comfast adaptation here has kept this config partition, why should I? There is nothing in there that could not be restored by returning to stock. It is just not needed and waste of space.

rhamp replied:
Please have alook [here](https://github.com/fall513/openwrt/blob/ar71xx-add-support-for-COMFAST-CF-EW71/EW72/target/linux/ar71xx/image/generic.mk). It is a fork by Ding Tengfei (dtf@comfast.cn) adding ar71xx support for different devices that have the config partition in comfast-fw. I also do have an EW72 and an E380ACv2, config part removed in openwrt, since restored when flashing back.

adrianschmutzler replied:
> No Comfast adaptation here has kept this config partition, why should I?

That's not true. CF-E560AC does (as I've stated before BTW).

@joaohcca Can you verify that the config partition is recreated with sensible data when you overwrite it in OpenWrt? Please save the contents beforehand and then actually overwrite the partition in OpenWrt (zeros or random) before you flash back, so we can be sure that it's recreated. Maybe you can even compare the backup with the new partition then.

joaohcca replied:
@adrianschmutzler I could try later I don't have the device with me right now. 
The DTS file I used on my PR doesn't override this partition. I can use dd and hexdump to compare the contents.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -130,12 +130,14 @@ case "$FIRMWARE" in
 			/lib/firmware/ath10k/QCA9888/hw2.0/board.bin
-		rm /lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin
+		board2file="/lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin"
+		[ -e $board2file ] && rm $board2file
```
Why do you think this is required?

rhamp replied:
Just to avoid error log entries in case there is no such file. It's just cleaner and doesn't hurt.

adrianschmutzler replied:
I can read that, but why do you expect that the existence of the file is random? If there is no file there, you wouldn't need the rm at all. So, please check whether you need to remove the file or not.
What you've done here is actually a workaround for not having to look ...

rhamp replied:
OK, if you don't want it, I take it out.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -330,6 +330,17 @@ define Device/comfast_cf-wr650ac-v2
+  DEVICE_VENDOR := COMFAST
+  DEVICE_MODEL := CF-WR752AC
+  DEVICE_VARIANT := v1
+DEVICE_PACKAGES := kmod-leds-gpio kmod-usb2 kmod-ath10k-ct ath10k-firmware-qca9888-ct \
```
Wrong indent, more than 80 characters per line.

rhamp replied:
OK, such rules are new to me.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -330,6 +330,17 @@ define Device/comfast_cf-wr650ac-v2
+  DEVICE_MODEL := CF-WR752AC
+  DEVICE_VARIANT := v1
+DEVICE_PACKAGES := kmod-leds-gpio kmod-usb2 kmod-ath10k-ct ath10k-firmware-qca9888-ct \
+	-uboot-envtools swconfig
```
Still haven't addressed my swconfig-related comment.

rhamp replied:
swconfig is not there by default, I had to add it. It is also used in cf-fw when in repeater mode (eth port is on phy4 of the switch, by hardware separated from the rest). But since there are other preset modi on firmware (as AP, router), it is up to the user to configure it. I can't replicate whole Comfast scripts.

adrianschmutzler replied:
There it is: https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/Makefile#L16

rhamp replied:
OK, I must have confounded with another platform. As said, I made that a few months ago. I also did  remove the kmod-leds-gpio, LEDs do work without too. But as far as I remember, taking out USB disables the PCIe (your comment from yesterday).

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -330,6 +330,17 @@ define Device/comfast_cf-wr650ac-v2
+  DEVICE_VARIANT := v1
+DEVICE_PACKAGES := kmod-leds-gpio kmod-usb2 kmod-ath10k-ct ath10k-firmware-qca9888-ct \
+	-uboot-envtools swconfig
+  IMAGE_SIZE := 16192k
```
Reminder: Needs to be changed if config partition is added.

rhamp replied:
see above

adrianschmutzler commented:
Regarding LEDs: I see that at least the match of GPIOs and color is consistent to #2652 now.
Still, the assignment of colors to use is entirely different.

@joaohcca Did you do a mistake there?
@rhamp Can you please state explicitly that LEDs are working correctly now (tested) and match behavior in vendor firmware.

adrianschmutzler commented:
> Partitioning is OK

@rhamp This is an example of you trying to get out of a discussion by denying that it exists. There are several other cases in this pull request. If you want your device supported here, and somebody raises concern, please properly address this concern and try to argue why your solution/implementation is better/correct/desirable.
You cannot get around this, and the only result you will earn from just stating "it is correct now" will be suspicion. Please explain and justify what you do.

joaohcca commented:
> Regarding LEDs: I see that at least the match of GPIOs and color is consistent to #2652 now.
> Still, the assignment of colors to use is entirely different.
> @joaohcca Did you do a mistake there?

I was following information provided by the hardware manufacturer, they sent me a file with the GPIO definitions:

- blue  GPIO12  低电平Low Electrical Level
- green GPIO15  低电平Low Electrical Level
- red GPIO11 低电平Low Electrical Level
- 按键reset  GPIO17  低电平Low Electrical Level
- 网口wan  P4   
 
The LED behavior from original firmware is to use green and red for signal indicators (strong and weak signal respectively) and blue for router/ap mode ref: [device page](http://en.comfast.com.cn/index.php?m=content&c=index&a=show&catid=14&id=50).

My PR led behavior it's different from original firmware, in fact this was one of the things I had to fix before finishing the port.

If you guys want to I can compile the image and flash the device later to run some tests later.

rhamp commented:
@adrianschmutzler

As **joaohcca** said, the LEDs on the device are dependent on the selected mode (AP/router/repeater). In AP mode, Comfast made a very long proprietary shell script that periodically checks the RSSI and makes only usage of 2 leds (either/or, not at the same time). So these aren't "classical " rssileds where multiple leds are powered simultaniouly, And these diodes are not used in the rest of the modes.

AFAIK Openwrt is missing ready to go LED scripts that clone this behavior. If someone has the time and desire anytime, it can be added.

But I think the approach joaohcca and I made is pretty standard for Openwrt, and better for "DIY people", that are ready to invest time (sometimes a lot as you know) in a custom Openwrt setup. Anyway, if someone doesn't like it, he can make an adaption.

It took me a lot of time to study the 9531 datasheet, to find out the pinmux stuff and get the leds to work. And by the way, the "rainbow effect" looks nice  when all are powered. ;-)

rhamp commented:
Strange, I made a forced push and files are not anymore showing up here.
They are still here https://github.com/openwrt/openwrt/compare/master...rhamp:cf-wr752ac

rhamp commented:
@adrianschmutzler
To convince you that the config partition is recreated at boot by stock firmware, here the dmesg output of one of my WR752AC-v1 (it was previously on my dts version without config partition). It was flashed using the u-boot web recovery, the only thing I changed at this point is shadow to get full root access.


_[   0.000000] Linux version 3.10.44 (qianhaichou@oracle.tecmint.com) (gcc version 4.8.3 (OpenWrt/Linaro GCC 4.8-2014.04 unknown) ) #11 Wed Sep 19 11:34:51 CST 2018
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU revision is: 00019374 (MIPS 24Kc)
[    0.000000] SoC: Qualcomm Atheros QCA9531 ver 2 rev 0
[    0.000000] Clocks: CPU:650.000MHz, DDR:391.333MHz, AHB:216.666MHz, Ref:25.000MHz
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x00000000-0x07ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x00000000-0x07ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000] free_area_init_node: node 0, pgdat 80352480, node_mem_map 81000000
[    0.000000]   Normal zone: 256 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 32512
[    0.000000] Kernel command line:  board=COMFAST-CF-WR752AC console=ttyS0,115200 mtdparts=spi0.0:64k(u-boot)ro,64k(art)ro,1536k(kernel),14592k(rootfs),64k(configs),64k(nvram)ro,16128k@0x20000(firmware) rootfstype=squashfs,jffs2 noinitrd
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 125796k/131072k available (2426k kernel code, 5276k reserved, 689k data, 276k init, 0k highmem)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:51
[    0.050000] Calibrating delay loop... 432.53 BogoMIPS (lpj=2162688)
[    0.050000] pid_max: default: 32768 minimum: 301
[    0.050000] Mount-cache hash table entries: 512
[    0.060000] NET: Registered protocol family 16
[    0.070000] MIPS: machine is COMFAST CF-WR752AC
[    0.510000] registering PCI controller with io_map_base unset
[    0.520000] bio: create slab <bio-0> at 0
[    0.530000] PCI host bridge to bus 0000:00
[    0.530000] pci_bus 0000:00: root bus resource [mem 0x10000000-0x11ffffff]
[    0.540000] pci_bus 0000:00: root bus resource [io  0x0000]
[    0.540000] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.550000] pci 0000:00:00.0: [168c:0056] type 00 class 0x028000
[    0.550000] pci 0000:00:00.0: invalid calibration data
[    0.550000] pci 0000:00:00.0: reg 10: [mem 0x00000000-0x001fffff 64bit]
[    0.550000] pci 0000:00:00.0: PME# supported from D0 D3hot
[    0.550000] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 00
[    0.550000] pci 0000:00:00.0: BAR 0: assigned [mem 0x10000000-0x101fffff 64bit]
[    0.560000] pci 0000:00:00.0: using irq 40 for pin 1
[    0.560000] Switching to clocksource MIPS
[    0.570000] NET: Registered protocol family 2
[    0.570000] TCP established hash table entries: 1024 (order: 1, 8192 bytes)
[    0.580000] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.580000] TCP: Hash tables configured (established 1024 bind 1024)
[    0.590000] TCP: reno registered
[    0.590000] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.600000] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.600000] NET: Registered protocol family 1
[    0.610000] PCI: CLS 0 bytes, default 32
[    0.620000] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.630000] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.640000] msgmni has been set to 245
[    0.650000] io scheduler noop registered
[    0.650000] io scheduler deadline registered (default)
[    0.650000] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.680000] serial8250.0: ttyS0 at MMIO 0x18020000 (irq = 11) is a 16550A
[    0.690000] console [ttyS0] enabled, bootconsole disabled
[    0.700000] ath79-spi ath79-spi: master is unqueued, this is deprecated
[    0.710000] m25p80 spi0.0: found w25q128, expected m25p80
[    0.720000] m25p80 spi0.0: w25q128 (16384 Kbytes)
[    0.720000] 7 cmdlinepart partitions found on MTD device spi0.0
[    0.730000] Creating 7 MTD partitions on "spi0.0":
[    0.730000] 0x000000000000-0x000000010000 : "u-boot"
[    0.740000] 0x000000010000-0x000000020000 : "art"
[    0.750000] 0x000000020000-0x0000001a0000 : "kernel"
[    0.750000] 0x0000001a0000-0x000000fe0000 : "rootfs"
[    0.760000] mtd: device 3 (rootfs) set to be root filesystem
[    0.770000] 1 squashfs-split partitions found on MTD device rootfs
[    0.770000] 0x000000780000-0x000000fe0000 : "rootfs_data"
[    0.780000] 0x000000fe0000-0x000000ff0000 : "configs"
[    0.790000] 0x000000ff0000-0x000001000000 : "nvram"
[    0.800000] 0x000000020000-0x000000fe0000 : "firmware"
[    0.820000] libphy: ag71xx_mdio: probed
[    0.830000] eth0: Atheros AG71xx at 0xba000000, irq 5
[    1.390000] ag71xx-mdio.1: Found an AR934X built-in switch
[    1.430000] eth1: Atheros AG71xx at 0xb9000000, irq 4
[    1.990000] ag71xx ag71xx.0: connected to PHY at ag71xx-mdio.1:04 [uid=004dd042, driver=Generic PHY]
[    2.000000] TCP: cubic registered
[    2.000000] NET: Registered protocol family 17
[    2.010000] Bridge firewalling registered
[    2.010000] 8021q: 802.1Q VLAN Support v1.8
[    2.020000] VFS: Mounted root (squashfs filesystem) readonly on device 31:3.
[    2.030000] Freeing unused kernel memory: 276K (8036b000 - 803b0000)
[    2.990000] init: Console is alive
[    2.990000] init: - watchdog -
[    5.010000] init: - preinit -
[    8.610000] jffs2: notice: (378) jffs2_build_xattr_subsystem: complete building xattr subsystem, 2 of xdatum (2 unchecked, 0 orphan) and 131 of xref (0 dead, 119 orphan) found.
[    8.630000] mount_root: switching to jffs2 overlay
[    9.080000] procd: - early -
[    9.090000] procd: - watchdog -
[    9.650000] procd: - ubus -
[   10.680000] procd: - init -
[   12.700000] natsemi dp8381x driver, version 2.1, Sept 11, 2006
[   12.700000]   originally by Donald Becker <becker@scyld.com>
[   12.700000]   2.4.x kernel port by Jeff Garzik, Tjeerd Mulder
[   12.730000] gre: GRE over IPv4 demultiplexor driver
[   12.750000] ip_gre: GRE over IPv4 tunneling driver
[   12.760000] i2c /dev entries driver
[   12.770000] Loading modules backported from Linux version master-2014-05-22-0-gf2032ea
[   12.780000] Backport generated by backports.git backports-20140320-37-g5c33da0
[   13.010000] PCI: Enabling device 0000:00:00.0 (0000 -> 0002)
[   13.020000] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   13.950000] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
[   13.960000] ath10k_pci 0000:00:00.0: kconfig debug 1 debugfs 1 tracing 1 dfs 0 testmode 1
[   13.980000] ath10k_pci 0000:00:00.0: firmware ver 10.4-3.5.1-00014 api 5 features no-p2p,mfp,peer-flow-ctrl,allows-mesh-bcast crc32 983b7b7a
[   14.370000] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id 0:0 crc32 4ea1ec87
[   16.170000] ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal otp max-sta 512 raw 0 hwcrypto 1
[   16.260000] ath: EEPROM regdomain: 0x0
[   16.260000] ath: EEPROM indicates default country code should be used
[   16.260000] ath: doing EEPROM country->regdmn map search
[   16.260000] ath: country maps to regdmn code: 0x3a
[   16.260000] ath: Country alpha2 being used: US
[   16.260000] ath: Regpair used: 0x3a
[   16.320000] u32 classifier
[   16.320000]     input device check on
[   16.320000]     Actions configured
[   16.340000] Mirror/redirect action on
[   16.370000] nf_conntrack version 0.5.0 (1969 buckets, 7876 max)
[   16.450000] Ebtables v2.0 registered
[   16.460000] ip_tables: (C) 2000-2006 Netfilter Core Team
[   16.520000] Netfilter messages via NETLINK v0.30.
[   16.580000] xt_time: kernel timezone is -0000
[   16.610000] ctnetlink v0.93: registering with nfnetlink.
[   16.620000] PPP generic driver version 2.4.2
[   16.630000] PPP MPPE Compression module registered
[   16.640000] NET: Registered protocol family 24
[   16.670000] ath: EEPROM regdomain: 0x0
[   16.670000] ath: EEPROM indicates default country code should be used
[   16.670000] ath: doing EEPROM country->regdmn map search
[   16.670000] ath: country maps to regdmn code: 0x3a
[   16.670000] ath: Country alpha2 being used: US
[   16.670000] ath: Regpair used: 0x3a
[   16.680000] ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[   16.700000] ieee80211 phy1: Atheros AR9531 Rev:2 mem=0xb8100000, irq=47
[   27.380000] device eth0 entered promiscuous mode
[   27.390000] device eth1 entered promiscuous mode
[   28.230000] heart_beat uses obsolete (PF_INET,SOCK_PACKET)
[   28.310000] br-lan: port 2(eth1) entered disabled state
[   29.100000] ath: EEPROM regdomain: 0x809c
[   29.100000] ath: EEPROM indicates we should expect a country code
[   29.100000] ath: doing EEPROM country->regdmn map search
[   29.100000] ath: country maps to regdmn code: 0x52
[   29.100000] ath: Country alpha2 being used: CN
[   29.100000] ath: Regpair used: 0x52
[   29.100000] ath: regdomain 0x809c dynamically updated by user
[   29.100000] ath: EEPROM regdomain: 0x809c
[   29.100000] ath: EEPROM indicates we should expect a country code
[   29.100000] ath: doing EEPROM country->regdmn map search
[   29.100000] ath: country maps to regdmn code: 0x52
[   29.100000] ath: Country alpha2 being used: CN
[   29.100000] ath: Regpair used: 0x52
[   29.100000] ath: regdomain 0x809c dynamically updated by user
..._

If l look under mtd5 (to be sure it is 5)...

_root@COMFAST:/# cat /sys/devices/platform/ath79-spi/spi_master/spi0/spi0.0/mtd/mtd5/name
configs_

if I "cat" mtd5, I get [this file](http://s000.tinyupload.com/?file_id=07755864896117815210
) ...

The file contains data up to 0x443E, the rest is filled up with FF.
I have still an WR752AC-**V2** with untouched original fw. Its configs partition is 256K long, but data ends at 0x443E too, the whole rest are FF. They seem to be  structurally resembling. cmp finds small differences (but it is also another, newer device type.)

The "restocked" device works normally, allows resetting to factory values. I can't find anything not working.
Also, since the V2 has only 8MB flash, 256K is a lot if you want to use enhancements, it is not a good precedent to force keeping the configs partition ...

Hoping this may convince you.

adrianschmutzler commented:
Okay, but the question is not whether the partitioning will be reestablished, but whether the _content_ is reestablished. For that, you have to alter the content of 0xfe0000+ under OpenWrt, then flash back to stock and look whether your changes are undone.

rhamp commented:
Maybe I do misunderstand you, but it seems you think that it could be a remnent that has never been overwritten?

adrianschmutzler commented:
I just want to know whether stock firmware writes back the content to this partition after OpenWrt might have altered it. That's quite easy to test as I explained above.

rhamp commented:
It seems mtd5 (configs) is not write protected in the stock fw. Is it OK for you if I just fill it with a regular pattern, and afterwards do check if the pattern is well present in 5?

rhamp commented:
And then the stock flashing again?

rhamp commented:
I made the regular pattern and mtd'ed it to mtd5, checked if it was well written, after a simple boot, I got a "nice" boot-log entry with my regular pattern:

_[   16.590000] PPP MPPE Compression module registered
[   16.590000] NET: Registered protocol family 24
[   16.650000] ieee80211 phy1: Atheros AR9531 Rev:2 mem=0xb8100000, irq=47
[   26.870000] device eth0 entered promiscuous mode
[   26.950000] device eth1 entered promiscuous mode
[   28.820000] heart_beat uses obsolete (PF_INET,SOCK_PACKET)
configs:do check configs
configs:revert size:1709rev_bak_size_str md5:"3DUfw▒▒▒▒▒▒▒▒"3DUfw▒▒▒▒▒▒▒▒"3DUfw▒▒▒▒▒▒▒▒"3DUfw▒▒▒▒▒▒▒▒ confis_md5:1709rev_bak_md5sum_str_now_

So it seems the pattern writing was successful, and the numbers in the first line of mtd5 seem to be an md5. Now I check if maybe mtd5 was auto repaired ... 

Yes, for me it looks like the configs check above also initiated new mtd5 config.

The new config mtd5 is 1:1 the same up to byte 0x3208 from that up to byte 0x41E5 some differences.

Now I reflash the stock to see if maybe it will recreate again a different mtd5.

Immediately after flashing stock bootlog showing:
_[   61.280000] heart_beat uses obsolete (PF_INET,SOCK_PACKET)
configs:/etc/changeconfig not found
configs:do update configs_

replacing again shadow for root ... and reboot 
... no more config bootlog entries at this time.

... that's nice, new mtd5 again differs slightly from the pre-flash version in the latter bytes.

My interpretation: it is recreated dynamically after each flash operation. up to byte 0x3208 seems to stay the same.

Checking if everything works ... wlans OK, ... web config by browser ... ok
... macs still have that strange form ;-)

rhamp commented:
**@adrianschmutzler**

what's your opinion ?

rhamp commented:
> I just want to know whether stock firmware writes back the content to this partition after OpenWrt might have altered it. That's quite easy to test as I explained above.

So to summarize my above findings, in Openwrt the address range 0xfe0000-0xfeffff is filled with with FF. I flashed stock, and configs partition is recreated and if it is empty, boot process rebuilds structure with each time the same content on absolute positions 0xfe0000-0xfe3208, the remaining bytes in the range 0xfe3209-0xfe41E5 seem to have partly a dynamic component. The rest is filled with FF up to the NVRAM partition, Everything seems to work well after stock fw flash.

HenryGreen2013 commented:
Thank you all a lot for the effort! 
May I ask if we could expect the support for CF-WR752AC from Openwrt soon?

rhamp commented:
Hi,
it is working as it is now, and you can always return to stock in case of problems. Anyway, a 2nd test user is needed to confirm it is working. Just add the changes manually to your files and post your results here if it is working for you too. Maybe adrianschmutzler will consider to add it then to official fw.

giutrec commented:
hi, i have that router, today I will compile the firmware from your repo and do a test.
I'll report you any problems.
However before anything I do a u-boot backup of the original firmware :)

rhamp commented:
Hi, thank you for testing and reporting any issues.

giutrec commented:
I have buold openwrt, but loading the sysupgrade image from the web panel does not bring any change

rhamp commented:
Did you try it using uboot http, or comfast fw http?

giutrec commented:
I've use comfast http

rhamp commented:
Can you try using uboot (push button, power up and hold for ~10sec, then http at 192.168.1.1)?

rhamp commented:
Are you sure it is a v1 and not a v2 (completely different hardware) ?

giutrec commented:
I don't know, i actually have a comfast clone marked hosome cf-wr752ac.
I think it's v1, but today I proceed to dismantle the repeater and see what hardware it is.

rhamp commented:
Google gives me this Hosome on German Amazon, 
https://www.amazon.de/Hosome-Repeater-Verst%C3%A4rker-unterst%C3%BCtzt-kompatibel/dp/B07QR1JJTT?th=1, 
and it says the chipset used is ..." Chip-Lösung: RTL8196EU + RTL8192ER ". This Openwrt commit is for Qualcomm QCA9531 + QCA9886 (Atheros). I have also made a commit for v2, that has Mediatek Chipset (Ramips) MT7628 + MT7612. So unfortunately, these commits don't seem to be for this clone.

adrianschmutzler commented:
So looks like it's no clone, and that's why different devices need different images ...

adrianschmutzler commented:
If you want this reviewed, you need to squash it into one proper commit.

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,134 @@
+};
+
+&pinmux {
+	led_wlan2g_pin: pinmux_wlan2g {
```
pinmux_wlan2g_led: wlan2g_led
and accordingly for 5g below

rhamp replied:
change done

adrianschmutzler commented on target/linux/ath79/dts/qca9531_comfast_cf-wr752ac-v1.dts:
```diff
@@ -0,0 +1,134 @@
+};
+
+&eth1 {
+	// "Disable" unpopulated GMAC1
```
Remove that comment.

rhamp replied:
removed

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -80,6 +80,9 @@ comfast,cf-e560ac)
 	;;
+comfast,cf-wr752ac-v1)
+  ucidef_set_led_netdev "lan" "LAN" "$boardname:blue:lan" "eth0"
+  ;;
```
Wrong indent, despite this can be merged with ecb1750 ...

rhamp replied:
merged with ecb1750

joaohcca commented:
Sorry for not replying earlier to this. I've flashed a binary with @rhamp code's. I can guarantee that you are able to rollback to COMFAST's firmware without affecting default behavior.

About the configs partition content, I'm afraid I can't help much in this discussion. My Original firmware it's different than the one he used. I didn't have the option to download a config back up to change the shadow and gain access to the device via ssh/serial port. I did upgrade device firmware via bootloader to version 2.3.0.2  to be able to run tests.

From the original firmware I've seen messages such as 
`configs:do check configs
[   54.380000] br-lan: port 2(eth1) entered disabled state
configs:revert size:1693rev_bak_size_str md5:���������������������������������������������������������������� confis_md5:1693rev_bak_md5sum_str_now
`
I'm not sure how the original firmware uses this partition, but I've seen the values of the final offset from the output of hexdump /dev/mtd5 varying even among comfast firmware's for this device. I wasn't able to find a pattern that I've edited at the end of the rootfs_data from rhamp's code after reflashing original firmare, but I think this wasn't the most effective way to tests this.

I can run more tests but it be nice to receive some instructions

rhamp commented:
Thank you joaohcca for making the test. Do you can confirm that my openwrt commit also works normally in daily usage (since a 2nd review/confirmation is needed to get it accepted)?

Are you sure you don't have the option to backup comfast fw settings? It creates a file called "bakup.file" that you have to rename to *.tar.gz. There should be a tool icon in the upper right with more settings, amongst others the backup option. I used the latest fw clownloaded from comfast.cn.

I assume the mtd5 config partition is some kind of file system. When recreated, it sets new time stamps, explaining the few bytes differences here and there. The unused rest of the partition is filled up with garbage.

Also a small change to my previous commit. I added a line to 02_network since default setting is 2 eth ports, but the device only has 1.

I'm hoping for a new evaluation by @adrianschmutzler.

joaohcca commented:
> Thank you joaohcca for making the test. Do you can confirm that my openwrt commit also works normally in daily usage (since a 2nd review/confirmation is needed to get it accepted)?
I did test the following: wireless radios, device LEDS and ethernet port. The device was able to provide a dhcp lease for my pc and receive an ip from my main router. I've managed to connect stations on both radios and browse web briefly. I've also been able to toogle leds using "echo 1(or 0) > /sys/class/leds/cf-wr752ac...

I'd recommend @rhamp  to change label = "cf-wr752ac:red:wlan5g"; to label = "cf-wr752ac**-v1**:red:wlan5g"; to match other leds labels.

> Are you sure you don't have the option to backup comfast fw settings? It creates a file called "bakup.file" that you have to rename to *.tar.gz. There should be a tool icon in the upper right with more settings, amongst others the backup option. I used the latest fw downloaded from comfast.cn.

This device was at the office for a few years, so I'm assuming it was flashed with older firmware versions from comfast. After updating to version CF-WR752ACV2-V2.3.0.5 I was able to export the bakup.file and access the terminal. 
 
> Also a small change to my previous commit. I added a line to 02_network since default setting is 2 eth ports, but the device only has 1.

I will update the code and report later, I believe that @adrianschmutzler will also ask you to remove the merge commits for a proper review. 

I' recommend you to avoid updating master branch from github web page to avoid the merge #1 commits.

Let know if you guys want me to run more tests.

rhamp committed:
ath79: add support for Comfast cf-wr752ac-v1

- Qualcomm QCA9531 + QCA9886
- dual band, antenna 2*3dBi
- Output power 50mW (17dBm)
- 1x 10/100 Mbps LAN RJ45
- 128 MB RAM / 16 MB FLASH (w25q128)
- 3 LEDs (red/green/blue)
  incorporated in
  "color wheel reset switch"
- UART 115200 8N1

Flashing instructions:

 The U-boot bootloader contains a recovery HTTP server
 to upload the  firmware. Push the reset button while powering the
 device on and keep it pressed for ~10 seconds. The device's LEDs will
 blink several times and the recovery page will be at
 http://192.168.1.1; use it to upload the sysupgrade image.

 Alternatively, the original firmware is based on OpenWrt so a
 sysupgrade image can be installed via the stock web GUI. Settings from
 the original firmware will be saved and restored on the new one, so a
 factory reset will be needed. To do so, once the new firmware is flashed,
 enter into failsafe mode by pressing the reset button several times during
 the boot process, until it starts flashing. Once in failsafe mode, perform
 a factory reset as usual.

LED-Info:

 The LEDs on the Comfast stock fw have a very proprietary behaviour,
 corresponding to the user selected working mode (AP, ROUTER or REPEATER).
 In the first two cases, only blue is used for status an LAN signaling. When
 using the latter, blue is always off (except for sysupgrade), either red
 signals bad rssi on master-link, or green good. Since there are no
 predefined working modes on OpenWRT, and usual RSSIleds are not EITHER/OR
 but AND, I left a quite default setting for blue for status and lan,
 green for wlan2g activity and red for wlan5g activity.

Signed-off-by: Roman Hampel <rhamp@arcor.de>

rhamp commented:
> I'd recommend @rhamp to change label = "cf-wr752ac:red:wlan5g"; to label = "cf-wr752ac**-v1**:red:wlan5g"; to match other leds labels.

Thanks, overseen this one. I've also removed the merge commits and replaced a false indent.

joaohcca commented:
> I've also removed the merge commits and replaced a false indent.
Check your git log. Try to pull the code from your branch but I'm seeing  a merge commit in there

commit bc434aed86d5d1ba99070644a29b2011cad4a49d
Merge: a4439e2 a57fcd8
Author: rhamp <60673606+rhamp@users.noreply.github.com>
Date:   Tue Mar 17 20:45:20 2020 +0100

    Merge pull request #10 from openwrt/master
    
    update
and It's causing merge errors when I pull

rhamp commented:
I have deleted these older merge commits in order to get a new single commit (and rebased to current release). New single commit is [b680449](https://github.com/rhamp/openwrt/commit/b6804493d7e259558f20ea4b4cab1b5c4cb257b4).

adrianschmutzler commented:
Hi, I've reviewed this PR another time and have merged an updated version into my staging tree here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

Note that I
- fixed the wmac MAC address based on the scheme evolved in the old PR
- changed LED names and behavior to resemble Router mode as described into commit message

If the vendor doesn't show Wifi activity in this case, we should as well _by default_. Anything else can be configured.

Furthermore:
@rhamp @joaohcca

Since this has been a joined effort, I'd like to change the authorship to reflect that. I've taken the PR from rhamp, but actually joaohcca was doing the same thing earlier, so he should receive credit as well. Thus, I'd propose the following:

```
Signed-off-by: Roman Hampel <rhamp@arcor.de>
Co-developed-by: Joao Albuquerque <joaohccalbu@gmail.com>
Signed-off-by: Joao Albuquerque <joaohccalbu@gmail.com>
... and then my stuff
```

I need explicit approval from both of you for adding that, so please reply whether you are okay with that or object.

joaohcca commented:
> fixed the wmac MAC address based on the scheme evolved in the old PR



> I need explicit approval from both of you for adding that, so please reply whether you are okay with that or object.

I'm ok with the changes, glad to see the device support moving forward.

rhamp commented:
No problem for me.

adrianschmutzler commented:
Thanks, merged as discussed.
